wdl-analysis/src/handlers/hover.rs

```diff
@@ -0,0 +1,335 @@
+        return Ok(Some(content));
+    }
+
+    // Finds hover information based on the scope.
+    if let Some(scope) = document.find_scope_by_position(token.span().start()) {
+        if let Some(name) = scope.lookup(token.text()) {
+            let kind = match name.ty() {
+                Type::Call(_) => "call",
+                _ => "variable",
+            };
+            return Ok(Some(format!(
+                "```wdl\n({kind}) {}: {}\n```",
+                token.text(),
+                name.ty()
+            )));
+        }
+    }
+
+    // Finds hover information across global definitions.
+    if let Some(content) = find_global_hover_in_doc(document, token)? {
+        return Ok(Some(content));
+    }
+
+    for (_, ns) in document.namespaces() {
+        // SAFETY: we know `get_index` will return `Some` as `ns.source` comes from
+        // `document.namespaces` which only contains namespaces for documents that
+        // are guaranteed to be present in the graph.
+        let node = graph.get(graph.get_index(ns.source()).unwrap());
+        let Some(imported_doc) = node.document() else {
+            continue;
+        };
+        if let Some(content) = find_global_hover_in_doc(imported_doc, token)? {
+            return Ok(Some(content));
+        }
+    }
+
+    Ok(None)
+}
+
+/// Resolves hover information based on the CST of the document.
+///
+/// This inspects the parent [`SyntaxNode`] of the hovered token to
+/// determine the most specific context.
+fn resolve_hover_by_context(
+    parent_node: &SyntaxNode,
+    token: &SyntaxToken,
+    document: &Document,
+    graph: &DocumentGraph,
+) -> Result<Option<String>> {
+    match parent_node.kind() {
+        SyntaxKind::TypeRefNode | SyntaxKind::LiteralStructNode => {
+            if let Some(s) = document.struct_by_name(token.text()) {
+                let root = if let Some(ns_name) = s.namespace() {
+                    // SAFETY: we just found a struct with this namespace name and the document
+                    // guarantees that `document.namespaces` contains a corresponding entry for
+                    // `ns_name`.
+                    let ns = document.namespace(ns_name).unwrap();
+                    let node = graph.get(graph.get_index(ns.source()).unwrap());
+                    node.document().unwrap().root()
+                } else {
+                    document.root()
+                };
+                return Ok(provide_struct_documentation(s, &root));
+            }
+        }
+        SyntaxKind::CallTargetNode => {
+            let target = CallTarget::cast(parent_node.clone()).unwrap();
+            let mut target_names = target.names();
+
+            let (callee_name, ns_name) = match (target_names.next(), target_names.next()) {
+                // Namespaced call
+                (Some(ns), Some(name)) => {
+                    if token.span() == name.span() {
+                        (name, Some(ns))
```

Comment:
Can we flip the order of the returns here to match the order we get them from `.next()`? It's a little confusing when reading the code to see them returned in the opposite order they came off the iterator.